cli/flox-activations/src/cli/start_or_attach.rs

```diff
@@ -22,113 +25,404 @@ pub struct StartOrAttachArgs {
 /// In either case, the activation can likely just be restarted.
-// TODO: use store_path rather than activation_id.
-// If we are activation 3, activation 1 fails, but activation 2 succeeds, we may want to attach to activation 2
 fn wait_for_activation_ready_and_attach_pid(
     activations_json_path: &Path,
-    activation_id: uuid::Uuid,
+    store_path: &str,
     attach_expiration: OffsetDateTime,
     attaching_pid: u32,
 ) -> Result<(), anyhow::Error> {
     loop {
-        let (activations, lock) = activations::read_activations_json(activations_json_path)?;
-        let Some(mut activations) = activations else {
-            anyhow::bail!("Expected an existing activations.json file");
+        let ready = check_for_activation_ready_and_attach_pid(
+            activations_json_path,
+            store_path,
+            attaching_pid,
+            attach_expiration,
+            OffsetDateTime::now_utc(),
+        )?;
+
+        if ready {
+            break;
+        }
+
+        std::thread::sleep(std::time::Duration::from_millis(100));
+    }
+    Ok(())
+}
+
+fn check_for_activation_ready_and_attach_pid(
+    activations_json_path: &Path,
+    store_path: &str,
+    attaching_pid: u32,
+    attach_expiration: OffsetDateTime,
+    now: OffsetDateTime,
+) -> Result<bool, anyhow::Error> {
+    let (activations, lock) = activations::read_activations_json(activations_json_path)?;
+    let Some(mut activations) = activations else {
+        anyhow::bail!("Expected an existing activations.json file");
+    };
+
+    let activation = activations
+        .activation_for_store_path_mut(store_path)
+        .context(indoc! {"
+            Prior activation of the environment completed.
+
+            Try again to start a new activation of the environment.
+        "})?;
+
+    if activation.ready() {
+        activation.attach_pid(attaching_pid, None);
+        activations::write_activations_json(&activations, activations_json_path, lock)?;
+        return Ok(true);
+    }
+
+    if !activation.startup_process_running() {
+        // TODO: clean out old activation of store_path
+        // Or we may need to do that in activation_for_store_path()
+        //
+        // TODO: just call StartOrAttach::handle again
+        anyhow::bail!(indoc! {"
+            Prior activation of the environment failed to start, or completed.
+
+            Try again to start a new activation of the environment.
+        "});
+    }
+
+    if now > attach_expiration {
+        anyhow::bail!(indoc! {"
+            Timed out waiting for a prior activation of the environment
+            to complete startup hooks.
+
+            Try again after the previous activation of the environment has completed.
+        "});
+    }
+    Ok(false)
+}
+
+#[cfg(test)]
+mod tests {
+    use uuid::Uuid;
+
+    use super::*;
+    use crate::cli::test::{read_activations, write_activations};
+
+    #[test]
+    fn attach_if_activation_exists() {
```

Comment:
nonblocking: I think these first two tests might actually be easier to reason about and give us better coverage if you don't pass in functions instead of `attach()` and `start()`

Reply:
Im not entirely sure what you ask for here, i assumed passing actual `attach` and `start` instead of `| .. | Ok(id)` functions and added that in [4c24a36](https://github.com/flox/flox/pull/2222/commits/4c24a36172a4ace89c5b7b64abd289dfde9f272e)

But at least for the attach case we'd now also actually be waiting for a process.
My intend with those two tests was to test that higher level function follows the right logic, i think testing the behaviour in integration with another process can be tests of its own and possibly target a lower level function altogether e.g. `check_for_activation_ready_and_attach_pid`.

I guess to have _full_ control over pid running states we'd need to pass a "check_pid" function / impl down to at least `Activation::startup_process_running`.
You mentioned we may have the preconditions for that merged now https://github.com/flox/flox/pull/2195#discussion_r1792450837.

But for the sake of size, I'd rather do that in a follow up

Reply:
Yeah I meant passing the actual `start` and `attach` functions